File: spk/cli/_cmd_export.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import Any
import os
import argparse

import structlog
from colorama import Fore

from . import _flags

import spk


_LOGGER = structlog.get_logger("spk.cli")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    export_cmd = sub_parsers.add_parser(
        "export", help=_export.__doc__, description=_export.__doc__, **parser_args
    )
    export_cmd.add_argument("package", metavar="PKG", help="The package to export")
    export_cmd.add_argument(
        "filename",
        metavar="FILE",
        nargs="?",
        help="The file to export into (Defaults to the name and verison of the package)",
    )
    export_cmd.set_defaults(func=_export)
    return export_cmd


def _export(args: argparse.Namespace) -> None:
    """Export a package as a tar file."""

    pkg = _flags.parse_idents(args.package)[0]
    build = ""
    if pkg.build is not None:
        build = f"_{pkg.build.digest}"
    if args.filename:
        filename = args.filename
    else:
        filename = f"{pkg.name}_{pkg.version}{build}.spk"
    try:
        spk.export_package(pkg, filename)
    except spk.storage.PackageNotFoundError:
        os.remove(filename)
        raise
    print(f"{Fore.GREEN}Created: {Fore.RESET}" + filename)

File: src/cli/cmd_export.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use anyhow::Result;
use clap::Args;
use colored::Colorize;

use super::flags;

/// Export a package as a tar file
#[derive(Args)]
pub struct Export {
    #[clap(flatten)]
    pub requests: flags::Requests,

    #[clap(short, long, global = true, parse(from_occurrences))]
    pub verbose: u32,

    /// The package to export
    #[clap(name = "PKG")]
    pub package: String,

    /// The file to export into (Defaults to the name and version of the package)
    #[clap(name = "FILE")]
    pub filename: Option<std::path::PathBuf>,
}

impl Export {
    pub fn run(&self) -> Result<i32> {
        let pkg = self
            .requests
            .parse_idents([self.package.as_str()])?
            .pop()
            .unwrap();

        let mut build = String::new();
        if let Some(b) = &pkg.build {
            build = format!("_{b}");
        }
        let filename = self.filename.clone().unwrap_or_else(|| {
            std::path::PathBuf::from(format!("{}_{}{build}.spk", pkg.name(), pkg.version))
        });
        let res = spk::storage::export_package(&pkg, &filename);
        if let Err(spk::Error::PackageNotFoundError(_)) = res {
            tracing::warn!("Ensure that you are specifying at least a package and");
            tracing::warn!("version number when exporting from the local repository");
        }
        if res.is_err() {
            if let Err(err) = std::fs::remove_file(&filename) {
                tracing::warn!(?err, path=?filename, "failed to clean up incomplete archive");
            }
        }
        res?;
        println!("{}: {:?}", "Created".green(), filename);
        Ok(0)
    }
}
